### PR TITLE
Prop types implementation per RFC

### PR DESCRIPTION
**Pull request by ibgreen**
This PR provides an initial implementation of the [Prop Types RFC](https://github.com/uber/deck.gl/blob/master/dev-docs/RFCs/v6.0/view-class-rfc.md)

### CODE REVIEW

File: examples/react/jsx-layers/README.md [hunk 1]
```diff
@@ -1,4 +0,0 @@
-# JSX-layers
```
Comment (ibgreen) [inline examples/react/jsx-layers/README.md]: This change is part of the JSX layers PR, better land that one first and rebase...

**Comment by ibgreen** [comment]
Bump: While there is an important open API audit question on the name of the "static field" containing the new prop types (i.e. `defaultProps` vs something else), the rest of the code in this PR should be reviewable. 

Great if I can get some reviews/opinions and get this ready to land on master now that 5.1 has branched out.

File: src/core/lifecycle/prop-types.js [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+  }
+  return {propTypes, defaultProps};
+}
+
+// Parses one property definition entry. Either contains:
+// * a valid prop type object ({type, ...})
+// * or just a default value, in which case type and name inference is used
+function parsePropType(name, propDef) {
+  switch (getTypeOf(propDef)) {
+    case 'object':
```
Comment (Pessimistress) [inline src/core/lifecycle/prop-types.js]: Nit: use enum?

Comment (ibgreen) [reply src/core/lifecycle/prop-types.js]: >Nit: use enum?

Yes, I'd normally go with an enum, but here with one or two exceptions, these are the JavaScript built in `typeof` return values. Not clear to me that aliasing them with an enum makes it easier to read. 

The main reason is that in that case it feels like there should there be a mapping TO the enum as well in the `getTypeOf` function, which might be overkill.

File: src/core/lifecycle/prop-types.js [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+  const {typeValidator} = typeDefinition;
+  if (typeValidator) {
+    assert(typeValidator(propDef), 'Illegal prop type');
+  }
+
+  return propDef;
+}
+
+function guessNumberType(name, value) {
+  const isKnownProp =
```
[the remark this answers is not in the document]
Comment (ibgreen) [reply src/core/lifecycle/prop-types.js]: Yes this is copied from the layer-browser, trying to keep that working without having to define all prop types. Agree that this should be removed. Will do in subsequent PRs.

File: src/core/lifecycle/prop-types.js [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    /radius|scale|width|height|pixel|size|miter/i.test(name) && /^((?!scale).)*$/.test(name);
+  const max = isKnownProp ? 100 : 1;
```
[the remark this answers is not in the document]
Comment (ibgreen) [reply src/core/lifecycle/prop-types.js]: Again copied from layer browser will be phased out once layers have actual type defs.

**Pull request merged**